lib/watchpool/watchpool.go

```diff
@@ -0,0 +1,169 @@
+// goroutine overhead, but beyond that it spawns ceiling(N/32) goroutines. For a
+// watch set containing 2048 radix nodes (e.g. ~682 service instances with
+// checks and nodes) that means each blocking query is running 64 goroutines on
+// top of the one actually serving the request. With hundreds or thousands of
+// clients watching the same things, this quickly adds up and places a lot of
+// load on the server. See https://github.com/hashicorp/consul/issues/4984 for
+// more info.
+//
+// By allowing different blocking queries that are watching the same radix nodes
+// to share a WatchSet.Watch, we can limit that number so in the example above,
+// one client still uses 64 goroutines to watch, but now 1000 clients can all be
+// blocking on that same query and still only 64 goroutines are needed (actually
+// 65 since we add a constant one more here per WatchSet).
+//
+// In the case where only a single RPC is blocked on a specific set of radix
+// nodes, we add no extra goroutines of overhead - the "leader" goroutine
+// processes the Watch just as before inline, if it times out with no change
+// observed, then it passes the leadership to one other watcher if any have come
+// along.
+//
+// To use it, just build your memdb.WatchSet like normal, but instead of calling
+// Watch or WatchCtx directly, call `pool.Watch(ctx, ws)`.
+type WatchPool struct {
+	sync.Mutex
+	watchSets map[uint64]*wsEntry
+}
+
+type wsEntry struct {
+	ws memdb.WatchSet
+	// done is closed by the leader Watch when any chan in the WatchSet is closed.
+	done chan struct{}
+	// leaderDone is NOT closed, but has a message sent on it if the current
+	// leader times out and returns to it's caller. If there are "follower" Watch
+	// calls, one of them will receive the message and take over being the leader.
+	leaderDone chan struct{}
+	// refs is a reference count for how many outstanding Watch calls there are
+	// for this WatchSet. It is incremented and decremented atomically.
+	refs uint32
+}
+
+// Watch performs a shared watch on a memdb.WatchSet. It will deduplicate calls
+// with the same WatchSets (i.e. exactly the same nodes in the state store being
+// watched) and run only one ws.WatchCtx call regardless of the number of
+// concurrent callers.
+func (p *WatchPool) Watch(ctx context.Context, ws memdb.WatchSet) error {
+	key, err := p.wsKey(ws)
+	if err != nil {
+		// This shouldn't ever really happen but if we return immediately then it
+		// turns whatever bug causes it into a blocking query hot loop. Instead,
+		// revert to normal un-pooled blocking execution.
+		return ws.WatchCtx(ctx)
+	}
+
+	// Lazy initialize the map
+	if p.watchSets == nil {
+		p.watchSets = make(map[uint64]*wsEntry)
+	}
+
+	// Whatever happens, we need to come back and decrement the ref count and
+	// possibly clean up at the end.
+	defer p.watcherDone(key)
+
+	// See if the key already exists in the share map
+	p.Lock()
```

Comment:
I didn't because RW locks are often not more efficient and my rule of thumb is generally to use simple locks unless the read/write ratio is very obviously unbalanced (like you write it only when reloading config) or I have realistic benchmarks to show it's actually better.

In this case it's moot because every caller here has to write - the leader creates the entry, the non-leader updates the reference count.

I also considered using atomic reference counts too but they are much harder to reason about whether the atomic increment and the locked reads and writes guarantee the ordering we actually need. Again if we had benchmarks to prove it's worth the effort to get right I'd reconsider but I don't want to prematurely optimize.